### PR TITLE
[fx] make sure args/kwargs are immutable

### PR DESCRIPTION
**Pull request by zdevito**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #45193 [fx] allow custom behavior for args, kwargs, and bool
* **#46121 [fx] make sure args/kwargs are immutable**

Otherwise, mutating them would make the uses/users lists inaccurate.
You can still mutate the node by assigning a new value to .args or .kwargs

Differential Revision: [D24232288](https://our.internmc.facebook.com/intern/diff/D24232288)

### CODE REVIEW

File: torch/fx/immutable_collections.py [hunk 1]
```diff
@@ -0,0 +1,13 @@
+def _no_mutation(self, *args, **kwargs):
+    raise NotImplementedError(f"'{type(self).__name__}' object does not support mutation")
```
Comment (jamesr66a) [inline torch/fx/immutable_collections.py:2]: Can we tell the user to explicitly assign new `args` and `kwargs`? Just so it's obvious what do you if you've gotten here

**Review by jamesr66a — APPROVED** [review]
image

**Comment by robieta** [comment]
This appears to have broken MyPy tests. (Observed in https://github.com/pytorch/pytorch/pull/46023) [CI Failure](https://fburl.com/k2v1s6jy)

**Comment by mruberry** [comment]
Unlanding. This broke multiple builds, as @robieta says. Relevant log snippet:

```
test_run_mypy - TestTypeHints
test_type_hints.py

Traceback (most recent call last):
  File "test_type_hints.py", line 217, in test_run_mypy
    self.fail(f"mypy failed: {stdout} {stderr}")
AssertionError: mypy failed: torch/fx/node.py:36: error: Need type annotation for '_uses' (hint: "_uses: Dict[<type>, <type>] = ...")  [var-annotated]
torch/fx/node.py:37: error: Argument 1 to "_update_args_kwargs" of "Node" has incompatible type "Union[Tuple[Any, ...], List[Any], Dict[str, Any], slice, Node, str, int, float, bool, dtype, Tensor, None]"; expected "Tuple[Union[Tuple[Any, ...], List[Any], Dict[str, Any], slice, Node, str, int, float, bool, dtype, Tensor, None], ...]"  [arg-type]
torch/fx/node.py:37: error: Argument 2 to "_update_args_kwargs" of "Node" has incompatible type "Union[Tuple[Any, ...], List[Any], Dict[str, Any], slice, Node, str, int, float, bool, dtype, Tensor, None]"; expected "Dict[str, Union[Tuple[Any, ...], List[Any], Dict[str, Any], slice, Node, str, int, float, bool, dtype, Tensor, None]]"  [arg-type]
torch/fx/node.py:103: error: Argument 1 to "_update_args_kwargs" of "Node" has incompatible type "Union[Tuple[Any, ...], List[Any], Dict[str, Any], slice, Node, str, int, float, bool, dtype, Tensor, None]"; expected "Tuple[Union[Tuple[Any, ...], List[Any], Dict[str, Any], slice, Node, str, int, float, bool, dtype, Tensor, None], ...]"  [arg-type]
torch/fx/node.py:111: error: Argument 2 to "_update_args_kwargs" of "Node" has incompatible type "Union[Tuple[Any, ...], List[Any], Dict[str, Any], slice, Node, str, int, float, bool, dtype, Tensor, None]"; expected "Dict[str, Union[Tuple[Any, ...], List[Any], Dict[str, Any], slice, Node, str, int, float, bool, dtype, Tensor, None]]"  [arg-type]
Found 5 errors in 1 file (checked 1099 source files)
```

**Pull request closed without merging**